### PR TITLE
Remove assumptions about exception handlers

### PR DESCRIPTION
**Pull request by cristina-david**
This PR modifies the conversion of exception handlers such that we don't rely on the fact that the first instruction in the handler is ```astore```. The root of the problem is caused by the fact that, during the bytecode conversion, exception handlers need to be considered as potential successors of function calls. Thus, for cases where the functions do not throw but exit normally, the stack will contain their normal return value (or nothing for void functions), which contradicts the expectations of the exception handlers. Before, this was handled by making certain assumptions about the first instruction existent in an exception handler. These assumptions were not obeyed by user code causing diffblue/platform#603. 

After this PR, the problem described above is solved by taking the following steps before the conversion of an exception handler: 1. clear the stack; 2. push the exceptional return object corresponding to the current function. These objects are fully instrumented later at the goto program level. This way the handler can be converted without any assumptions as the right exception is on the stack (regardless of who its predecessor is).

### CODE REVIEW

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 1]
```diff
@@ -1305,48 +1306,37 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
+        auxiliary_symbolt new_symbol;
+        new_symbol.is_static_lifetime=true;
+        int file_name_length=
+          id2string(method.source_location.get_file()).size();
+        // remove the file extension
+        const std::string &file_name=
+          id2string(method.source_location.get_file()).
+             substr(0, file_name_length-5);
+        // generate the name of the exceptional return variable
+        const std::string &exceptional_var_name=
```
Comment (pkesseli) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: This creates the same exception variable name for two different methods using the following code: 

```java
class Ex extends Throwable {}

class A {
  public static void f() {
    try {
      throw new Ex();
    }
    catch(Ex ex) { }
  }
}

class B {
  public static void f() {
    try {
      throw new Ex();
    }
    catch(Ex ex) { }
  }
}

public class test {
 public static void main (String arg[]) {
   A.f();
   B.f();
 }
}
```

Perhaps consider making the `class_symbol` available to `convert_instructions` and use its name? E.g. pass it as a parameter or make it a field of `java_bytecode_convert_methodt`?

File: src/java_bytecode/java_entry_point.cpp [hunk 2]
```diff
@@ -589,14 +589,17 @@ bool java_entry_point(
...
 
-  // add the exceptional return value
-  auxiliary_symbolt exc_symbol;
-  exc_symbol.mode=ID_C;
-  exc_symbol.is_static_lifetime=false;
-  exc_symbol.name=id2string(symbol.name)+EXC_SUFFIX;
-  exc_symbol.base_name=id2string(symbol.name)+EXC_SUFFIX;
-  exc_symbol.type=typet(ID_pointer, empty_typet());
-  symbol_table.add(exc_symbol);
+  if(!symbol_table.has_symbol(id2string(symbol.name)+EXC_SUFFIX))
```
Comment (pkesseli) [inline src/java_bytecode/java_entry_point.cpp]: This generates a different symbol name than the catch handler in `java_bytecode_convert_method.cpp` if the main class does not have the same name as the file, e.g.:

```java
class Ex extends Throwable {}

class test2 {
  public static void main() {
   try {
     throw new Ex();
   }
   catch(Ex ex) { }
  }
}

public class test {
 public static void main (String arg[]) {
 }
}
```

When calling CBMC with --function test2.main, this will produce `"java::test2.main:()V#exception_value"` and `"java::test.main:()V#exception_value"` in `java_bytecode_convert_method.cpp`. This could probably be solved by using the class name in `java_bytecode_convert_method.cpp`.

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 3]
```diff
@@ -2407,6 +2397,10 @@ codet java_bytecode_convert_methodt::convert_instructions(
       results[1]=op[0];
       results[0]=op[1];
     }
+    else if(statement=="nop")
```
Comment (pkesseli) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Perhaps this could go into a separate pull request to master rather than test-gen-support? For this unit test we could just use some other dummy instructions (e.g. `iconst_0`). @mgudemann

File: src/java_bytecode/java_entry_point.cpp [hunk 4]
```diff
@@ -589,14 +589,17 @@ bool java_entry_point(
...
+  {
+    // add the exceptional return value
+    auxiliary_symbolt exc_symbol;
+    exc_symbol.mode=ID_C;
+    exc_symbol.is_static_lifetime=false;
```
Comment (pkesseli) [inline src/java_bytecode/java_entry_point.cpp]: Is there a reason why the symbol is not static in this case?

File: regression/cbmc-java/exceptions19/exception_handler.j [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        .limit stack 4
+        .limit locals 4
+
+      begin:
+        new java/lang/Exception
+        dup
+        invokespecial java/lang/Exception.<init>()V
+        athrow
+      start:
+        nop
```
Comment (mgudemann) [inline regression/cbmc-java/exceptions19/exception_handler.j]: @pkesseli is probably right, having someting like
```
start:
 iconst_0
 astore_1
end:
 astore1
...
```
is better, then `nop` integration could become a PR to master

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 6]
```diff
@@ -1305,48 +1306,37 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
-        handler_block.move_to_operands(catch_handler);
-        handler_block.move_to_operands(label_block);
-        c=handler_block;
-        break;
+        // at the beginning of a handler, clear the stack and
+        // push the corresponding exceptional return variable
+        stack.clear();
+        auxiliary_symbolt new_symbol;
+        new_symbol.is_static_lifetime=true;
+        int file_name_length=
```
Comment (mgudemann) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: use `size_t`

File: src/java_bytecode/java_bytecode_convert_method.cpp [hunk 7]
```diff
@@ -1305,48 +1306,37 @@ codet java_bytecode_convert_methodt::convert_instructions(
...
+          id2string(method.source_location.get_file()).size();
+        // remove the file extension
+        const std::string &file_name=
+          id2string(method.source_location.get_file()).
+             substr(0, file_name_length-5);
```
Comment (mgudemann) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: please describe this magic number (probably to remove `.java`?) or define `const string` with the value and get its length, could even be done as `constexpr` at compile time

**Comment by cristina-david** [comment]
@pkesseli You are right about the naming of the exception vars. However, this is a problem relating to the whole exception handling (vars created in goto-programs/remove-exceptions.cpp do not use the class id either) so I've opened an issue on this (diffblue/test-gen#498) and I'll fix it there.

**Comment by cristina-david** [comment]
@pkesseli I'm now using function id when generating the name of the exception var.

File: src/java_bytecode/java_entry_point.cpp [hunk 8]
```diff
@@ -589,14 +589,17 @@ bool java_entry_point(
...
-  exc_symbol.type=typet(ID_pointer, empty_typet());
-  symbol_table.add(exc_symbol);
+  if(!symbol_table.has_symbol(id2string(symbol.name)+EXC_SUFFIX))
+  {
+    // add the exceptional return value
+    auxiliary_symbolt exc_symbol;
+    exc_symbol.mode=ID_C;
+    exc_symbol.is_static_lifetime=true;
+    exc_symbol.name=id2string(symbol.name)+EXC_SUFFIX;
+    exc_symbol.base_name=id2string(symbol.name)+EXC_SUFFIX;
```
Comment (pkesseli) [inline src/java_bytecode/java_entry_point.cpp]: Maybe:

```C++
  exc_symbol.base_name=exc_symbol.name;
```
